### PR TITLE
Detect, report, and enable fix for resource quota violations

### PR DESCRIPTION
**Pull request by thallgren**
## Description

If a namespace has a resource quota, all containers of all pods in that namespace must specify their resources. Failing to do so, means that the pod cannot be created. This PR contains what's necessary to specify such resources for the injected traffic-agent and init-container.

The PR also contains logic to detect when an intended injection of the traffic-agent fails. This is done by watching cluster events, and filtering out the ones that is relevant for the injection. When errors occur, the user is informed about them and given an advice on how to move forward (typically adding resource definitions to the Helm chart).

In order to make the use-case "a) update helm chart with resources, b) run a helm upgrade", this PR also adds the logic necessary to regenerate the telepresence-agent config map. The regeneration takes place when the injector service starts,
and when SystemA changes its response to what traffic-agent image that should be used.

There's also some clean-up in the area of obtaining the traffic-agent image from SystemA that hopefully will offload that service quite a bit.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

### CODE REVIEW

**Comment by thallgren** [comment]
Closes #1806

File: cmd/traffic/cmd/agent/agent_test.go [hunk 1]
```diff
@@ -10,8 +10,8 @@ import (
 
 	"github.com/spf13/afero"
 	"github.com/stretchr/testify/require"
-	"gopkg.in/yaml.v3"
 	core "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
```
Comment (raphaelreyna) [inline cmd/traffic/cmd/agent/agent_test.go:14]: Why the swap in yaml dep?

File: cmd/traffic/cmd/manager/internal/state/intercept.go [hunk 2]
```diff
@@ -355,10 +341,68 @@ func waitForConfigMapUpdate(ctx context.Context, cmAPI typed.ConfigMapInterface,
...
-func (s *State) waitForAgent(ctx context.Context, name, namespace string) error {
+func watchFailedInjectionEvents(ctx context.Context, name, namespace string) (<-chan *events.Event, error) {
+	// A timestamp with second granularity is needed here, because that's what the event creation time uses.
+	// Finer granularity will result in relevant events seemingly being created before this timestamp because
+	// they have the fraction of seconds trimmed off (which is odd, given that the type used is a MicroTime).
+	start := time.Unix(time.Now().Unix(), 0)
+
+	ei := k8sapi.GetK8sInterface(ctx).EventsV1().Events(namespace)
+	w, err := ei.Watch(ctx, meta.ListOptions{
+		FieldSelector: fields.OneTermNotEqualSelector("type", "Normal").String()})
```
Comment (raphaelreyna) [inline cmd/traffic/cmd/manager/internal/state/intercept.go:352]: could this be further filtered down to only get events related to the agent?

File: cmd/traffic/cmd/manager/internal/state/intercept.go [hunk 3]
```diff
@@ -355,10 +341,68 @@ func waitForConfigMapUpdate(ctx context.Context, cmAPI typed.ConfigMapInterface,
...
+				msg := fe.Note
+				switch fe.Reason {
+				case "Unhealthy":
+					// Let readiness probe continue, this isn't fatal
+					continue
+				case "BackOff":
+					// The traffic-agent container was injected, but it fails to start
+					msg = fmt.Sprintf("%s\nThe logs of %s %s might provide more details", msg, fe.Regarding.Kind, fe.Regarding.Name)
+				case "FailedCreate", "FailedScheduling":
+					// The injection of the traffic-agent failed for some reason, most likely due to resource quota restrictions.
```
Comment (raphaelreyna) [inline cmd/traffic/cmd/manager/internal/state/intercept.go:399]: I feel like the "most likely" part of this comment should be reflected in the message sent to the user unless we are actually certain that a resource quota restriction was the cause of the issue.

**Review by raphaelreyna — APPROVED** [review]
lgtm; just left a suggestion regarding user message and a couple of questions

File: cmd/traffic/cmd/manager/internal/state/intercept.go (hunk 2 above)
Comment (thallgren) [reply cmd/traffic/cmd/manager/internal/state/intercept.go:352]: The filtering mechanism is very limited. It only allows for exact matches and only AND conditions can be used. That why I decided to just match abnormal events here and then do more advanced filtering further down in this method. It shouldn't matter much unless the cluster is in a very bad shape and sends a very large number of abnormal events.

File: cmd/traffic/cmd/manager/internal/state/intercept.go (hunk 3 above)
Comment (thallgren) [reply cmd/traffic/cmd/manager/internal/state/intercept.go:399]: It should. I made the hint a bit more descriptive.

File: cmd/traffic/cmd/agent/agent_test.go (hunk 1 above)
Comment (thallgren) [reply cmd/traffic/cmd/agent/agent_test.go:14]: It's well explained in the commit comment.

**Review by raphaelreyna — APPROVED** [review]
nice

**Pull request merged**